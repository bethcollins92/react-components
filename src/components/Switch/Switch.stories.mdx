import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Switch from "./Switch";

<Meta title="Switch" component={Switch} />

export const Template = (args) => <Switch {...args} />;

### Switch

This is a [React](https://reactjs.org/) component for the Vanilla [Switch](https://vanillaframework.io/docs/patterns/switch) component.

### Props

<ArgsTable of={Switch} />

### Default

<Canvas>
  <Story name="Default" args={{ label: "Switch" }}>
    {Template.bind({})}
  </Story>
</Canvas>

### Disabled

<Canvas>
  <Story name="Disabled" args={{ disabled: true, label: "Disabled switch" }}>
    {Template.bind({})}
  </Story>
</Canvas>
