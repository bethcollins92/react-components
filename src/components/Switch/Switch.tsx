import React from "react";
import type { HTMLProps, ReactNode } from "react";

import Field from "../Field";

import type { PropsWithSpread } from "types";

export type Props = PropsWithSpread<
  {
    /**
     * The label name for the switch
     */
    label: ReactNode;
    /**
     * Whether the switch is disabled or not
     */
    disabled?: boolean;
  },
  HTMLProps<HTMLInputElement>
>;

export const Switch = ({
  label,
  disabled = false,
  ...inputProps
}: Props): JSX.Element => {
  return (
    <Field>
      <label className="p-switch">
        <input
          type="checkbox"
          className="p-switch__input"
          role="switch"
          disabled={disabled}
          {...inputProps}
        ></input>
        <span className="p-switch__slider"></span>
        <span className="p-switch__label">{label}</span>
      </label>
    </Field>
  );
};

export default Switch;
