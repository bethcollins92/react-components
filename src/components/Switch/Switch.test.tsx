import { shallow } from "enzyme";
import React from "react";

import Switch from "./Switch";

describe("<Switch />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    const component = shallow(<Switch label="Switch" />);

    expect(component).toMatchSnapshot();
  });

  it("renders and matches the snapshot with disabled", () => {
    const component = shallow(<Switch label="Disabled switch" disabled />);

    expect(component).toMatchSnapshot();
  });

  //unit tests
  it("renders a switch with a label", () => {
    const component = shallow(<Switch label="Switch" />);

    expect(component.find("label").first().hasClass("p-switch")).toBeTruthy();
  });

  it("renders a disabled switch", () => {
    const component = shallow(<Switch label="Disabled switch" disabled />);

    expect(component.find("input[disabled]").length).toBe(1);
  });
});
