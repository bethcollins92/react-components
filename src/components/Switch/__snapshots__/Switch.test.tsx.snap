// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> renders and matches the snapshot 1`] = `
<label
  class="p-switch"
>
  <input
    class="p-switch__input"
    role="switch"
    type="checkbox"
  />
  <span
    class="p-switch__slider"
  />
  <span
    class="p-switch__label"
  >
    Switch
  </span>
</label>
`;

exports[`<Switch /> renders and matches the snapshot with disabled 1`] = `
<label
  class="p-switch"
>
  <input
    class="p-switch__input"
    disabled=""
    role="switch"
    type="checkbox"
  />
  <span
    class="p-switch__slider"
  />
  <span
    class="p-switch__label"
  >
    Disabled switch
  </span>
</label>
`;
